cli/testflinger_cli/__init__.py

```diff
@@ -883,36 +904,105 @@ def artifacts(self):
+            )
+
+        log_tuples = [
+            (phase_logs["last_fragment_number"], phase_logs["log_data"])
+            for phase in TestPhase
+            if (phase_logs := log_dict.get(phase.value))
+        ]
+        if not log_tuples:
+            return -1, ""
+        fragment_numbers, log_data_list = zip(*log_tuples, strict=False)
+        last_fragment_number = max(fragment_numbers)
+        combined_logs = "".join(log_data_list)
+        return last_fragment_number, combined_logs
+
+    def poll(self, log_type: LogType):
         """Poll for output from a job until it is completed."""
+        start_fragment = self.args.start_fragment
+        start_timestamp = self.args.start_timestamp
+        job_id = self.args.job_id
+        phase = self.args.phase
         if self.args.oneshot:
             # This could get an IOError for connection errors or timeouts
             # Raise it since it's not running continuously in this mode
-            output = self.get_latest_output(self.args.job_id)
-            if output:
-                print(output, end="", flush=True)
+            last_fragment_number, log_data = self._get_combined_log_output(
+                job_id, log_type, phase, start_fragment, start_timestamp
+            )
+
+            if last_fragment_number < 0:
+                print("Waiting on Output")
+            else:
+                print(log_data, end="", flush=True)
+                print(f"Last Fragment Number: {last_fragment_number}")
             sys.exit(0)
-        self.do_poll(self.args.job_id)
+        if self.args.json:
+            output_json = self.client.get_logs(
+                job_id,
+                log_type,
+                phase,
+                start_fragment,
+                start_timestamp,
+            )
+            print(json.dumps(output_json, indent=4))
+            sys.exit(0)
+
+        self.do_poll(job_id, log_type)
+
+    def poll_output(self):
+        """Poll for agent output from a job until it is completed."""
+        self.poll(LogType.STANDARD_OUTPUT)
 
     def poll_serial(self):
         """Poll for serial output from a job until it is completed."""
-        if self.args.oneshot:
-            output = self.get_latest_serial_output(self.args.job_id)
-            if output:
-                print(output, end="", flush=True)
-            sys.exit(0)
-        self.do_poll_serial(self.args.job_id)
+        self.poll(LogType.SERIAL_OUTPUT)
 
-    def do_poll(self, job_id):
+    def do_poll(
+        self,
+        job_id: str,
+        log_type: LogType = LogType.STANDARD_OUTPUT,
+    ):
         """Poll for output from a running job and print it while it runs.
 
         :param str job_id: Job ID
+        :param LogType log_type: Enum representing serial or agent output.
         """
+        start_fragment = self.args.start_fragment
+        start_timestamp = self.args.start_timestamp
+        job_id = self.args.job_id
+        phase = self.args.phase
```

Comment:
Based on my other comment I realized about the self.args! Submit/reserve don't have this arguments defined, one option could be to include it as params in the `do_poll` but I think this would work just fine:
```suggestion
        start_fragment = getattr(self.args, 'start_fragment', 0)
        start_timestamp = getattr(self.args, 'start_timestamp', None)
        phase = getattr(self.args, 'phase', None)
```

Please let me know what you think.